app/src/main/java/com/example/tasky/agenda/data/repository/OfflineFirstTaskRepository.kt

```diff
@@ -0,0 +1,45 @@
+package com.example.tasky.agenda.data.repository
+
+import com.example.tasky.agenda.domain.data.TaskRepository
+import com.example.tasky.agenda.domain.data.network.TaskRemoteDataSource
+import com.example.tasky.agenda.domain.model.Task
+import com.example.tasky.core.domain.data.TaskLocalDataStore
+import com.example.tasky.core.domain.util.DataError
+import com.example.tasky.core.domain.util.EmptyResult
+import com.example.tasky.core.domain.util.Result
+import com.example.tasky.core.domain.util.asEmptyDataResult
+
+class OfflineFirstTaskRepository(
+    private val taskRemoteDataSource: TaskRemoteDataSource,
+    private val taskLocalDataStore: TaskLocalDataStore,
+) : TaskRepository {
+
+    override suspend fun createTask(task: Task): EmptyResult<DataError> {
+        val localResult = taskLocalDataStore.upsertTask(task)
+        if (localResult !is Result.Success) {
+            return localResult.asEmptyDataResult()
+        }
+
+        val remoteResult = taskRemoteDataSource.createTask(task)
+        return when (remoteResult) {
+            is Result.Error -> {
```

Comment:
I've actually started to extend this class with these utilities:

```kotlin
inline fun <T, E: Error> Result<T, E>.onSuccess(action: (T) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> this
        is Result.Success -> {
            action(data)
            this
        }
    }
}
inline fun <T, E: Error> Result<T, E>.onError(action: (E) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> {
            action(error)
            this
        }
        is Result.Success -> this
    }
}
```

Much more convenient IMO:

```
repository.signUp(payload).onSuccess { responseData ->

}.onError { error -> 
    _state.update { it.copy(error = error.toUiText()) }
}
```